File: 04-Data-Structures/code/typescript/src/mutexMap.ts
import { DSError } from "./errors.js";
import { ChainingHashMap } from "./hashMap.js";

/**
 * Mutex-safe map API: every operation acquires a single exclusive lock
 * around the underlying map access. This lab runs single-threaded, so the
 * lock body is exercised sequentially (as documented in `bounded and mutex
 * concurrency vectors`); the guard still detects reentrancy bugs, which is
 * the property that matters when this same code is later run under real
 * concurrency.
 */
export class MutexMap<V> {
  private map = new ChainingHashMap<V>();
  private locked = false;

  private withLock<T>(fn: () => T): T {
    if (this.locked) throw new DSError("invalid", "reentrant lock acquisition detected");
    this.locked = true;
    try {
      return fn();
    } finally {
      this.locked = false;
    }
  }

  set(key: string, value: V): void {
    this.withLock(() => this.map.set(key, value));
  }

  get(key: string): V | null {
    return this.withLock(() => this.map.get(key));
  }

  delete(key: string): boolean {
    return this.withLock(() => this.map.delete(key));
  }

  size(): number {
    return this.withLock(() => this.map.size());
  }

  checkInvariants(): void {
    this.map.checkInvariants();
  }
}

File: 04-Data-Structures/code/python/seb_ds/mutex_map.py
"""Mutex-safe map API. Mirrors `typescript/src/mutexMap.ts`.

Every operation acquires a single exclusive lock around the underlying map
access. This lab runs single-threaded, so the lock body is exercised
sequentially; the guard still detects reentrancy bugs, which is the property
that matters when this same code is later run under real concurrency (a real
`threading.Lock` would be dropped in here unchanged).
"""

from __future__ import annotations

from typing import Callable, Generic, Optional, TypeVar

from .errors import DSError
from .hash_map import ChainingHashMap

V = TypeVar("V")
R = TypeVar("R")


class MutexMap(Generic[V]):
    def __init__(self) -> None:
        self._map: ChainingHashMap[V] = ChainingHashMap()
        self._locked = False

    def _with_lock(self, fn: Callable[[], R]) -> R:
        if self._locked:
            raise DSError("invalid", "reentrant lock acquisition detected")
        self._locked = True
        try:
            return fn()
        finally:
            self._locked = False

    def set(self, key: str, value: V) -> None:
        self._with_lock(lambda: self._map.set(key, value))

    def get(self, key: str) -> Optional[V]:
        return self._with_lock(lambda: self._map.get(key))

    def delete(self, key: str) -> bool:
        return self._with_lock(lambda: self._map.delete(key))

    def size(self) -> int:
        return self._with_lock(lambda: self._map.size())

    def check_invariants(self) -> None:
        self._map.check_invariants()
